### PR TITLE
feat: add mysql impl

### LINKED ISSUE #15: Add default database implementation
### Describe This Problem

Target of Sqlness is to ensure SQL's correction of one database, so if a database want to use sqlness, it need to implement `Database` trait, although this works as expected, this seems a tedious work.

In SQL world, there are only a few wire protocols, such as MySQL or PostgreSQL. If we can supply a default database with MySQL protocols, then any MySQL-compatible databases can use sqlness with less work.

### Proposal

Add one default database implementation: MySQL

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by dust1**
# Which issue does this PR close?

Closes #15 

# Rationale for this change
 

Added a default database implementation based on mysql

# What changes are included in this PR?

A default impl module has been added to append internal database implementations

# Are there any user-facing changes?

If the target database is mysql, users do not need to implement the `Database`.
I tried to consider the possibility of adding other database implementations, So you can use this method to create default implementations of various databases:
```
pub async fn build_default_database<T>(config: DatabaseConnConfig) -> Result<Box<dyn Database>>
where
    T: DatabaseBuilder,
{
    let builder = T::default();
    builder.build(config).await
}
```
Its name is now `_build_default_database` because I need to get it through `make clippy`.

If you want to add new database implementations, We can adapt this method by implementing `MysqlDatabaseBuilder` and `Database`

# How does this change test

I ran it through my local database and it worked:
`select * from test;`
```
id,name,age,
1,Bob,25,
2,Ben,41
```

`delete from test where id = 1`
```
affected_rows: 1
```

### CODE REVIEW

**Comment by waynexia** [comment]
I plan to review it today or tomorrow

**Comment by dust1** [comment]
> I plan to review it today or tomorrow

It's still a little bit unimplemented, but I'll submit it later or later in the evening😥

**Comment by waynexia** [comment]
> > I plan to review it today or tomorrow
> 
> It's still a little bit unimplemented, but I'll submit it later or later in the evening😥

Sorry I didn't notice this is a draft PR on my phone... Nooo means to push, take your own pace 🫣

**Comment by dust1** [comment]
Sorry it took so long😢

File: Cargo.toml [hunk 1]
```diff
@@ -10,9 +10,13 @@ repository = "https://github.com/CeresDB/sqlness"
 license = "Apache-2.0"
 description = "SQL integration test harness"
 
+[dev-dependencies]
+tokio-test = "*"
+
 [dependencies]
 async-trait = "0.1"
 derive_builder = "0.11"
```
Comment (jiacai2050) [inline Cargo.toml:15]: Make this dep optional, and add a feature to enable this.

File: Cargo.toml [hunk 2]
```diff
@@ -10,9 +10,13 @@ repository = "https://github.com/CeresDB/sqlness"
 license = "Apache-2.0"
```
Comment (jiacai2050) [inline Cargo.toml]: Delete Cargo.lock, it shouldn't be tracked in git.

https://github.com/CeresDB/sqlness/issues/28

File: src/config.rs [hunk 3]
```diff
@@ -69,3 +80,15 @@ impl Config {
         false
     }
 }
+
+impl Display for DatabaseConnConfig {
```
Comment (jiacai2050) [inline src/config.rs]: How about just `derive(Debug)`?

File: src/database_impl/setup.rs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+                } else {
+                    format!("{}", MysqlFormatter { rows })
+                }
+            }
+            Err(e) => format!("Failed to execute query, err: {:?}", e),
+        })
+    }
+}
+
+pub async fn _build_default_database<T>(config: DatabaseConnConfig) -> Result<Box<dyn Database>>
```
Comment (jiacai2050) [inline src/database_impl/setup.rs]: If this method is not used, I think we remove it.

**Review by jiacai2050 — CHANGES_REQUESTED** [review]

File: src/database_impl/setup.rs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    sync::{Arc, Mutex},
+};
+
+use async_trait::async_trait;
+use mysql::{prelude::Queryable, Conn, OptsBuilder, Row};
+
+use crate::{config::DatabaseConnConfig, error::Result, Database, SqlnessError};
+
+#[async_trait]
+pub trait DatabaseBuilder: Send + Sync + Default {
```
Comment (jiacai2050) [inline src/database_impl/setup.rs]: Why this trait depend on Default?

**Comment by jiacai2050** [comment]
@dust1 Please sign CLA first.

File: Cargo.toml (hunk 1 above)
Comment (dust1) [reply Cargo.toml:15]: it like this?
```
tokio-test = {version = "0.4.2", optional = true}
```

File: src/database_impl/setup.rs (hunk 4 above)
Comment (dust1) [reply src/database_impl/setup.rs]: This method is the one I used to create the default database implementation.
```rust
let mysql_db = build_default_database::<MysqlDatabaseBuilder>(config)?;
```
But right now sqlness doesn't need it anywhere

File: src/database_impl/setup.rs (hunk 5 above)
Comment (dust1) [reply src/database_impl/setup.rs]: As with `build_default_database` above, I need Default to be compatible with other database implementations

File: src/database_impl/setup.rs (hunk 4 above)
Comment (jiacai2050) [reply src/database_impl/setup.rs]: I have on clear idea on when will this be used, so I suggest remove this method for now.

File: src/config.rs (hunk 3 above)
Comment (jiacai2050) [reply src/config.rs]: If we have `Debug`, then we can remove this `Display` impl?

File: src/database_impl/setup.rs (hunk 5 above)
Comment (jiacai2050) [reply src/database_impl/setup.rs]: I still can't understand why this is required.

Mysql builder should have no relation with other DB.

File: src/database_impl/setup.rs [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    async fn build(&self, config: DatabaseConnConfig) -> Result<Box<dyn Database>>;
```
Comment (waynexia) [inline src/database_impl/setup.rs]: How about making `Database` (the return type) a type parameter, like
```rust
trait DatabaseBuilder {
    type Database: Database;
    
    async fn build(&self, config: DatabaseConnConfig) -> Result<Self::Database>;
}
```

As the type to be built is determined. And as well as the config type, maybe?

File: src/database_impl/setup.rs (hunk 5 above)
Comment (dust1) [reply src/database_impl/setup.rs]: That's what I think. If additional database support is added later, the Builder can be implemented like this:
```
pub struct PostgreSQLBuilder;

#[async_trait]
impl DatabaseBuilder for PostgreSQLBuilder {
    async fn build(&self, config: DatabaseConnConfig) -> Result<Box<dyn Database>> {
        todo!()
    }
}
```
We can then create a default implementation of PostgreSQL like this:
```
let pg_db = build_default_database::<PostgreSQLBuilder>()
```
Maybe I should delete them now😂

File: Cargo.toml (hunk 1 above)
Comment (jiacai2050) [reply Cargo.toml:15]: Yes, you also need a feature to enable this. Something like this
```
[features]
default = []
mysql = ["dep:mysql"]

```

File: src/config.rs (hunk 3 above)
Comment (dust1) [reply src/config.rs]: I need to print this configuration information in the error, which requires me to have `Debug` and implement `Display`
```rust
    #[error("Failed to connection database server. config: {config}, error: {source}")]
    ConnFailed {
        source: mysql::Error,
        config: DatabaseConnConfig,
    },
```

**Comment by jiacai2050** [comment]
The general impl looks fine to me now, I will append some minor changes before merge.

**Review by jiacai2050 — APPROVED** [review]
LGTM

**Pull request merged**